Tighten job validation when JobManagedBy is enabled

mimowo opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind documentation
/kind feature
/kind api-change
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/enhancements/issues/4368

#### Special notes for your reviewer:

The commit to review here: https://github.com/kubernetes/kubernetes/pull/125849/commits/afe83f70fd6fc931a142058b0c2069f6b1e2d3f2, since the first commit is basically https://github.com/kubernetes/kubernetes/pull/125510, so the PR will be rebased after merge. 

The validation was originally introduced but reverted in: https://github.com/kubernetes/kubernetes/pull/123792. I don't revert the API comment changes since the changes for validation are only in alpha for now. I think we could update the comments when graduating to Beta.

I've tested that with the new changes and validation rules added when the JobManagedBy is enabled in Beta (
https://github.com/kubernetes/kubernetes/pull/125816/commits/7e7706957d4e1b9547cca1dc481a0a9180995a69). then all e2e tests pass, and there are no validation failures during them (no `resp=422` in API server logs for jobs)

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Introduce the following validation rules for Job status when JobManagedBy is enabled:
1. the count of ready pods is less or equal than active
2. when transitioning to terminal phase for Job, the number of terminating pods is 0
3. terminal Job conditions (Failed and Complete) should be preceded by adding the corresponding interim conditions: FailureTarget and SuccessCriteriaMet
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4368-support-managed-by-for-batch-jobs
```

mimowo commented:
/test pull-kubernetes-integration 
Looks like overloaded infra: "There are no nodes that your pod can schedule to - check your requests, tolerations, and node selectors"

mimowo commented:
/assign @dejanzele @atiratree @alculquicondor 
Ready for review. This should conclude the effort for second alpha for JobManagedBy.

mimowo commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

mimowo committed:
Add more test cases for SuccessCriteriaMet

Cleanup error messages in the new code

Add validation for the Job controller fields

alculquicondor commented on pkg/apis/batch/validation/validation.go:
```diff
@@ -518,6 +518,16 @@ func validateJobStatus(job *batch.Job, fldPath *field.Path, opts JobStatusValida
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("completionTime"), status.CompletionTime, "completionTime cannot be set before startTime"))
 		}
 	}
+	if opts.RejectFailedJobWithoutFailureTarget {
```
I'm not sure we should be validating this (same for SuccessCriteriaMet).

You could imagine a job controller that doesn't have these conditions but still waits for everything to finish before setting the terminal condition.

mimowo replied:
Yeah, I'm not fully sure either, but this was the suggestion: https://github.com/kubernetes/enhancements/pull/4657#discussion_r1620853394. 

I think it is reasonable to assume that any Job controller implementation should communicate the fate of the Job as soon as possible, especially since now setting the terminal condition can be delayed. So, adding the validation we make sure the developers are aware of the conditions and would not introduce custom conditions for this purpose.

alculquicondor replied:
I would have preferred to save on complexity of validation code, but since it's already approved in the KEP, I'm ok with it.

deads2k replied:
It's also important for consumers of the Job API, that all the job controller implementations communicate status in the same way with the same meanings.  Accepting some complexity in validation to ensure that API consumers can be assured of congruent semantics between implementations is net win for the community.

alculquicondor commented on pkg/apis/batch/validation/validation.go:
```diff
@@ -568,6 +578,16 @@ func validateJobStatus(job *batch.Job, fldPath *field.Path, opts JobStatusValida
+		}
+	}
+	if opts.RejectMoreReadyThanActivePods {
+		if status.Ready != nil && *status.Ready > status.Active {
```
Is this now possible because terminating pods can't be counted as ready?

mimowo replied:
It is now possible after https://github.com/kubernetes/kubernetes/pull/125546. 

However, even before this PR terminating pods weren't counted as ready. This is because ready pods were counted as a subset of active pods (which exclude terminating), but only active was updated during the syncJob.

soltysh reviewed (approved):
/lgtm
/approve
/hold 
so we merge in the right order

soltysh commented:
/triage accepted
/priority important-longterm

mimowo commented:
/retest

mimowo commented:
/test pull-kubernetes-integration
overloaded infra

mimowo committed:
do not assert in e2e tests

mimowo committed:
Delay setting terminal Job conditions until all pods are terminal

Fix after rebase

# Conflicts:
#	pkg/controller/job/job_controller_test.go
#	test/e2e/apps/job.go
#	test/integration/job/job_test.go

mimowo committed:
Extend the e2e tests

mimowo committed:
Verify the Job status active, ready and terminating are 0

Signed-off-by: Michal Wozniak <michalwozniak@google.com>

# Conflicts:
#	test/e2e/apps/job.go

mimowo commented:
/close 
as https://github.com/kubernetes/kubernetes/pull/125510 is getting merged now with all the changes
